File: tests/test_geography.R
#!/usr/bin/env Rscript

here::i_am("code/paths.R")

source(here::here("code", "c00_shared", "geography.R"))

stopifnot(
  identical(state_fips(c(1, "06", 12)), c("01", "06", "12")),
  identical(county_code(c(1, "013", 101)), c("001", "013", "101")),
  identical(county_fips(c(1001, "06037")), c("01001", "06037")),
  identical(combine_county_fips(c(1, "06"), c(1, 37)), c("01001", "06037")),
  identical(state_from_county_fips(c("01001", 6037)), c("01", "06")),
  identical(county_code_from_county_fips(c("01001", 6037, "46102")),
    c("001", "037", "113")
  ),
  identical(harmonize_county_fips_2010(c("46102", "06037")),
    c("46113", "06037")
  ),
  identical(cz_id(c("00100", 200)), c("100", "200")),
  identical(aewr_region_id(c("01", 17)), c("1", "17")),
  identical(oews_area_code(c("0900001", 33100)), c("0900001", "33100"))
)

valid <- data.frame(
  state_fips = c("01", "06"),
  county_code = c("001", "037"),
  county_fips = c("01001", "06037"),
  cz_id = c("100", "200"),
  aewr_region_id = c("1", "17"),
  oews_area_code = c("0900001", "33100")
)
stopifnot(identical(assert_geo_columns(
  valid,
  names(valid)
), valid))

expect_error <- function(expression) {
  inherits(try(force(expression), silent = TRUE), "try-error")
}

stopifnot(
  expect_error(state_fips("123")),
  expect_error(county_fips("06A37")),
  expect_error(aewr_region_id("18")),
  expect_error(assert_geo_columns(
    transform(valid, county_fips = as.integer(county_fips)),
    "county_fips"
  )),
  expect_error(assert_geo_columns(
    transform(valid, county_fips = c(NA_character_, "06037")),
    "county_fips"
  )),
  expect_error(assert_geo_columns(
    transform(valid, county_fips = c("46102", "06037")),
    "county_fips"
  )),
  expect_error(assert_geo_columns(valid, "missing_geo_field"))
)

cat("geography tests passed\n")

File: tests/test_geography.py
"""Unit tests for the Python geographic-code contract."""

from __future__ import annotations

import polars as pl

from h2a.geography import (
    assert_geo_columns,
    county_code_from_county_fips,
    harmonize_county_fips_2010,
    normalize_county_fips_list,
    normalize_geo_code,
    state_from_county_fips,
)


def expect_error(function, *args, **kwargs) -> None:
    try:
        function(*args, **kwargs)
    except (TypeError, ValueError):
        return
    raise AssertionError("Expected geographic contract validation to fail")


def test_normalization() -> None:
    assert normalize_geo_code(1, "state_fips") == "01"
    assert normalize_geo_code(" 037 ", "county_code") == "037"
    assert normalize_geo_code(6037.0, "county_fips") == "06037"
    assert normalize_geo_code("00100", "cz_id") == "100"
    assert normalize_geo_code("01", "aewr_region_id") == "1"
    assert normalize_geo_code("0900001", "oews_area_code") == "0900001"
    assert harmonize_county_fips_2010("46102") == "46113"
    assert state_from_county_fips("06037") == "06"
    assert county_code_from_county_fips("06037") == "037"
    assert county_code_from_county_fips("46102") == "113"
    assert normalize_county_fips_list("06037, 46102") == "06037,46113"


def test_fail_fast_validation() -> None:
    expect_error(normalize_geo_code, "123", "state_fips")
    expect_error(normalize_geo_code, "06A37", "county_fips")
    expect_error(normalize_geo_code, "18", "aewr_region_id")

    valid = pl.DataFrame(
        {
            "state_fips": ["01", "06"],
            "county_fips": ["01001", "06037"],
            "cz_id": ["100", "200"],
        }
    )
    assert_geo_columns(valid, ["state_fips", "county_fips", "cz_id"])

    expect_error(
        assert_geo_columns,
        valid.with_columns(pl.col("county_fips").cast(pl.Int64)),
        ["county_fips"],
    )
    expect_error(
        assert_geo_columns,
        pl.DataFrame({"county_fips": [None, "06037"]}, schema={"county_fips": pl.String}),
        ["county_fips"],
    )
    expect_error(
        assert_geo_columns,
        pl.DataFrame({"county_fips": ["46102"]}),
        ["county_fips"],
    )
    expect_error(assert_geo_columns, valid, ["missing_geo_field"])


if __name__ == "__main__":
    test_normalization()
    test_fail_fast_validation()
    print("Python geography tests passed")
